### PR TITLE
[SYCL] Fix bit_cast for half type

### PR DESCRIPTION
**Pull request by maximdimakov**
After removing of half type trivial default constructor the bit_cast to sycl::half became impossible. Added helper struct for that case
Signed-off-by: mdimakov <maxim.dimakov@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/bit_cast.hpp [hunk 1]
```diff
@@ -20,6 +22,10 @@ namespace sycl {
 // forward decl
 namespace detail {
 inline void memcpy(void *Dst, const void *Src, std::size_t Size);
+
+template <typename Type> struct type_helper { using T = Type; };
+
+template <> struct type_helper<sycl::half> { using T = uint16_t; };
```
Comment (vladimirlaz) [inline sycl/include/CL/sycl/bit_cast.hpp:28]: half is special type and can lower to different types on host and device. Something similar to https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/detail/generic_type_traits.hpp#L398-L399 should be used.

File: sycl/include/CL/sycl/bit_cast.hpp [hunk 2]
```diff
@@ -20,6 +22,10 @@ namespace sycl {
 // forward decl
 namespace detail {
 inline void memcpy(void *Dst, const void *Src, std::size_t Size);
+
+template <typename Type> struct type_helper { using T = Type; };
```
Comment (vladimirlaz) [inline sycl/include/CL/sycl/bit_cast.hpp:26]: can we rename the struct to something more explaining? lowering_type? or something which sounds better but also explains what it does.

**Review by vladimirlaz — CHANGES_REQUESTED** [review]

File: sycl/include/CL/sycl/bit_cast.hpp [hunk 3]
```diff
@@ -8,6 +8,8 @@
 
 #pragma once
 
+#include <CL/sycl/aliases.hpp>
```
[the remark this answers is not in the document]
Comment (maximdimakov) [reply sycl/include/CL/sycl/bit_cast.hpp]: I removed this include

**Review by vladimirlaz — CHANGES_REQUESTED** [review]
Could you please also add regression test?

File: sycl/test/regression/bit_cast_lin.cpp [hunk 4]
```diff
@@ -0,0 +1,4 @@
+// RUN: %clangxx -fsycl -fsycl-host-compiler=g++ -fsycl-host-compiler-options='-std=c++17' %s -o %t.out
```
Comment (vladimirlaz) [inline sycl/test/regression/bit_cast_lin.cpp:1]: Why do we need to change host compiler?

Comment (maximdimakov) [reply sycl/test/regression/bit_cast_lin.cpp:1]: The problem was reproducible only if host compiler was g++

**Review by vladimirlaz — DISMISSED** [review]

**Review by vladimirlaz — APPROVED** [review]

**Pull request merged**